hub-saml/src/test/java/uk/gov/ida/saml/hub/validators/response/idp/IdpResponseValidatorTest.java

```diff
@@ -74,14 +88,37 @@ public void shouldValidateSamlResponseSignature() {
     @Test
     public void shouldValidateSamlAssertionSignature() {
         Assertion assertion = mock(Assertion.class);
-        List<Assertion> assertions = List.of(assertion);
+        List<Assertion> assertions = ImmutableList.of(assertion);
         ValidatedResponse validatedResponse = mock(ValidatedResponse.class);
-
+        Issuer responseIssuer = mock(Issuer.class);
+        
+        when(validatedResponse.getIssuer()).thenReturn(responseIssuer);
+        when(responseIssuer.getValue()).thenReturn("issuerValue");
         when(samlResponseSignatureValidator.validate(response, IDPSSODescriptor.DEFAULT_ELEMENT_NAME)).thenReturn(validatedResponse);
         when(assertionDecrypter.decryptAssertions(validatedResponse)).thenReturn(assertions);
-
+        when(badAssertionDecrypter.decryptAssertions(validatedResponse)).thenThrow(SamlFailedToDecryptException.class);
+        
         validator.validate(response);
 
         verify(samlAssertionsSignatureValidator).validate(assertions, IDPSSODescriptor.DEFAULT_ELEMENT_NAME);
     }
+
+    @Test
+    public void shouldThrowIfAllDecryptersFail() {
+        String issuerValue = "issuerValue";
+        ValidatedResponse validatedResponse = mock(ValidatedResponse.class);
+        Issuer responseIssuer = mock(Issuer.class);
+
+        when(validatedResponse.getIssuer()).thenReturn(responseIssuer);
+        when(responseIssuer.getValue()).thenReturn(issuerValue);
+        when(samlResponseSignatureValidator.validate(response, IDPSSODescriptor.DEFAULT_ELEMENT_NAME)).thenReturn(validatedResponse);
+        when(assertionDecrypter.decryptAssertions(validatedResponse)).thenThrow(SamlFailedToDecryptException.class);
+        when(badAssertionDecrypter.decryptAssertions(validatedResponse)).thenThrow(SamlFailedToDecryptException.class);
+
+        samlValidationException.expect(SamlFailedToDecryptException.class);
+        samlValidationException.expectMessage(String.format("Could not decrypt IDP assertions from %s with any of the decrypters", issuerValue));
+        validator.validate(response);
+        verify(idpDecryptionErrorCount).labels(validatedResponse.getIssuer().getValue());
+        verify(childCounter).inc();
```

Comment:
These two lines are never being run. The test catches the error thown by `validator.validate()`, handles it and then exits. It should be possible to use `assertThrows` rather than `ExpectedException` which apparently is part of junit 5. I've briefly tried to use it but intellij doesn't seem to find it, which is confusing as it looks like we specify `org.junit.jupiter:junit-jupiter-api:5.5.2` in our gradle file...

The mocks that they're verifying are also never actually used - this is why it would be nice to be pass the `Counter` in to the class that uses it rather than just newing one up - otherwise I don't think it's possible to stub it out.

Reply:
So I tried this, but it required a lot of changes and I couldn't get the integration test passing. So I have gone back to static counter, I know you don't like them

Reply:
OK, no worries. I think you can remove the `ExpectedException samlValidationException` rule from this class now you're using the `assertThrows` in the `shouldThrowIfAllDecryptersFail` test.